### DISCUSSION TITLE
BUG: How to update a subscription plan options, and remove previously selected toggle/checkbox option?

### DISCUSSION THREAD

**Discussion by Krislunde** [swellstores/community#300] [Help]
Hey, so, I'm trying to let customers edit their active subscription plans and change options that were made when the order was originally created. Changing set options (a "select" option field) to another value/choice works just fine (f.ex changing "size" from medium to small, is fine) however, the documentation is unclear on how to completely remove option items, such as a toggle/checkbox. 

I'm following the documentation on https://developers.swell.is/frontend-api/subscriptions - and using swell.subscriptions.update('subscription.id', { options })

The problem scenario: a customer might have checked a checkbox for "Please assemble" or "add cheese". How do we "unset" this option from a subscription that already has it? I had a look at how "edit subscription options" is implemented in the Origin storefront, and discovered that the productEditOptionsPopup component actually doesn't work at all, because the updatePlan() function on the subscriptions/id/edit.vue page actually submits the options object where the subscription id is supposed to go. I've fixed this, and now changing the value of fields works, like the aforementioned "size" from one value to another, but I can't figure out how to "remove" toggles/checkboxes from a plan. The documentation doesn't show an example. The origin implementation is to send an empty string, but that doesn't seem to work. I've tried with values like false, undefined, and removing the checkbox from the options object entirely when updating a subscription. Nothing works. The subscription seems to get updated, the _position_ of the checkbox field in the subscription object saved in swell changes, but the value is still there (with a value of the fields _Name_), and the price of the subscription is the same, and I can't figure out how to remove it. However, adding a checkbox/toggle to a subscription that didn't already have it works as expected.

Please advise.

**Comment by Krislunde** [comment]
I'm still unable to solve this. Does noone have any idea how to remove an active toggle option from a customer subscription, via swell-js? What value should the option be set to? It seems impossible to remove the option, or set it to false, or whatever. The consequence is that customers subscription may remain at a higher price, because new product variant is not chosen.

**Reply by ericingram** [reply]
@Krislunde try this (note I haven't recently tested this using the frontend API but it works on the backend API)

```
swell.subscriptions.update(id, {
  $pull: { 
    options: { 
      [id or name]: id_or_name_to_remove
    }
  }
})
```

**Reply by steven-k1** [reply]
@Krislunde Checking back to see if the suggestion made by @ericingram resolved the issue you were encountering.

**Reply by Krislunde** [reply]
My apologies! I had to pivot to finish a different project, but I will make sure to test this before the weekend. I will report back either way!

**Reply by steven-k1** [reply]
@Krislunde No worries at all. Just want to make sure the issue is resolved before it gets lost in the mix of things.

**Reply by Krislunde** [reply]
Unfortunately, this fix does not seem to work from the frontend API.

> PUT https://storeid.swell.store/api/subscriptions/64c08db104bae90011643c07 400
> Uncaught (in promise) Error: You are not allowed to update `$pull`

Even if it had worked, it would not have been an ideal fix.  Instead of merely running `await swell.subscriptions.update(subscription.value.id, options)` we would have to run a check on every option to see if they are a toggle, if they are empty, compare with the saved options on server, and selectively remove them using $pull, then update all the other options that might have been changed.

Not being able to remove toggles from active subscriptions effectively means they can not be used, especially if they cause price/product variant changes.

**Reply by sags95** [reply]
@Krislunde Would you be able to try this without the `$pull` param? You should be able to update the value for a toggle option like so:

```
    var update = await swell.subscriptions.update('<subscription_id>', {
      options: [
        {
          id: "<option_id>",
          value: false
        }
      ]
    });
```

**Reply by Krislunde** [reply]
Hi @sags95 !

Setting the option to "false" does nothing. On the other hand, setting the toggle option to "true" actually disables the toggle for some reason, whereas setting the value to the field name toggles it on. However, the toggle does not update in the swell admin to show the active status, but the correct variant is chosen -> $20 cost is added.

However, there are still issues. When inspecting the subscription in the Swell admin panel, the toggle option never actually visually turns on this way. See screenshots for clarification here: [Imgur album with pictures](https://imgur.com/a/jsLZKt4)

There seems to be a bug in Swell here, somewhere. Also, it's a bit of an annoying developer experience having to write checks for whether an option has changed, whether its a toggle, and look use its ID to change value instead of just passing an object with names and values of all options - as shown in the Origin storefront. As well as having to use "true" and "option name" instead of true/false or even option.name/false.

However, the biggest issue is that the Toggle option is not visually updated in Swell admin even though the correct variant is chosen.

**Reply by sags95** [reply]
Hey @Krislunde , we'll take a deeper look at this issue you're describing here. If there's a bug I'll be sure to file it with the team.

**Reply by Krislunde** [reply]
Thanks, good to hear. I'll keep an eye on this issue. :)

Good luck!

**Reply by Krislunde** [reply]
@logeshswell How come this is considered resolved? The issue still persists, or has there been changes made to the backend?

**Reply by logeshswell** [reply]
@Krislunde I didn't mark it as resolved/answered intentionally. I just closed the discussion and will reopen it back when there's an update.

Please feel free to reopen it back if you have further questions.

**Reply by Krislunde** [reply]
Roger, thanks :)
